pkg/resources/procedure.go

```diff
@@ -182,136 +186,368 @@ func Procedure() *schema.Resource {
+	args, diags := getProcedureArguments(d)
+	if diags != nil {
+		return diags
+	}
+	if len(args) > 0 {
+		req.WithArguments(args)
+	}
+
+	// read optional params
+	if v, ok := d.GetOk("execute_as"); ok {
+		if strings.ToUpper(v.(string)) == "OWNER" {
+			req.WithExecuteAs(sdk.Pointer(sdk.ExecuteAsOwner))
+		} else if strings.ToUpper(v.(string)) == "CALLER" {
+			req.WithExecuteAs(sdk.Pointer(sdk.ExecuteAsCaller))
+		}
 	}
+	if v, ok := d.GetOk("null_input_behavior"); ok {
+		req.WithNullInputBehavior(sdk.Pointer(sdk.NullInputBehavior(v.(string))))
+	}
+	if v, ok := d.GetOk("comment"); ok {
+		req.WithComment(sdk.String(v.(string)))
+	}
+	if v, ok := d.GetOk("secure"); ok {
+		req.WithSecure(sdk.Bool(v.(bool)))
+	}
+
+	if err := client.Procedures.CreateForSQL(ctx, req); err != nil {
+		return diag.FromErr(err)
+	}
+	argTypes := make([]sdk.DataType, 0, len(args))
+	for _, item := range args {
+		argTypes = append(argTypes, item.ArgDataType)
+	}
+	sid := sdk.NewSchemaObjectIdentifierWithArguments(database, schema, name, argTypes)
+	d.SetId(sid.FullyQualifiedName())
+	return ReadContextProcedure(ctx, d, meta)
+}
 
-	if err := snowflake.Exec(db, q); err != nil {
-		return fmt.Errorf("error creating procedure %v err = %w", name, err)
+func createPythonProcedure(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	name := d.Get("name").(string)
+	schema := d.Get("schema").(string)
+	database := d.Get("database").(string)
+	id := sdk.NewSchemaObjectIdentifier(database, schema, name)
+
+	returns, diags := parseProcedureReturnsRequest(d.Get("return_type").(string))
+	if diags != nil {
+		return diags
+	}
+	procedureDefinition := d.Get("statement").(string)
+	runtimeVersion := d.Get("runtime_version").(string)
+	packages := []sdk.ProcedurePackageRequest{}
+	for _, item := range d.Get("packages").([]interface{}) {
+		packages = append(packages, *sdk.NewProcedurePackageRequest(item.(string)))
+	}
+	handler := d.Get("handler").(string)
+	req := sdk.NewCreateForPythonProcedureRequest(id, *returns, runtimeVersion, packages, handler)
+	req.WithProcedureDefinition(sdk.String(procedureDefinition))
+	args, diags := getProcedureArguments(d)
+	if diags != nil {
+		return diags
+	}
+	if len(args) > 0 {
+		req.WithArguments(args)
 	}
 
-	procedureID := &procedureID{
-		DatabaseName:  database,
-		SchemaName:    schema,
-		ProcedureName: name,
-		ArgTypes:      builder.ArgTypes(),
+	// read optional params
+	if v, ok := d.GetOk("execute_as"); ok {
+		if strings.ToUpper(v.(string)) == "OWNER" {
+			req.WithExecuteAs(sdk.Pointer(sdk.ExecuteAsOwner))
+		} else if strings.ToUpper(v.(string)) == "CALLER" {
+			req.WithExecuteAs(sdk.Pointer(sdk.ExecuteAsCaller))
+		}
 	}
 
-	d.SetId(procedureID.String())
+	// TODO: [ { CALLED ON NULL INPUT | { RETURNS NULL ON NULL INPUT | STRICT } } ] does not work for java, scala or python
```

Comment:
I did not notice this TODO earlier: can we add issue number to it?

Reply:
this isn't something we can fix. i did post in the docs-discuss channel though https://snowflake.slack.com/archives/C6380540P/p1707511734666249.

Reply:
also updated this comment to link to discussion in docs-discuss channel